README.md

```diff
@@ -1,2 +1,254 @@
+
+The conformance validation tool is primarily intended to be run on a specified version of Linkerd to verify the correctness of a [Kubernetes](https://kubernetes.io/) cluster's configuration with respect to Linkerd as well as validate non-trivial network communication (HTTP, gRPC, websocket) among stateless and stateful workloads in the Linkerd data plane.
+
+The conformance tests exercise the following features:
+
+- [ ] Validation of your Linkerd2 control plane
+- [ ] Automatic proxy injection on workloads
+- [ ] Functioning of  `linkerd tap`, `stat`, `routes` and `edges` commands
+- [ ] Verifying the functioning of the `tap` extension API server
+- [ ] Retries and timeouts
+- [ ] Data plane health checks
+- [ ] Ingress configuration
+
+_...and a lot more to come_
+
+The Linkerd project is hosted by the Cloud Native Computing Foundation ([CNCF](https://www.cncf.io/)).
+
+## Table of Contents
+
+- [Repository Structure](https://github.com/mayankshah1607/linkerd2-conformance#repository-structure)
+- [Configuring your tests](https://github.com/mayankshah1607/linkerd2-conformance#configuring-your-tests)
+  - [Configuration Options](https://github.com/mayankshah1607/linkerd2-conformance#configuration-options)
+- [Usage](https://github.com/mayankshah1607/linkerd2-conformance#usage)
+  - [Using the Sonobuoy CLI](https://github.com/mayankshah1607/linkerd2-conformance#using-the-sonobuoy-cli)
+  - [Running the tests using Docker]()
+  - [Running the tests locally](https://github.com/mayankshah1607/linkerd2-conformance#running-the-tests-locally)
+- [Adding new tests](https://github.com/mayankshah1607/linkerd2-conformance#adding-new-tests)
+  - [Initial bootstrapping](https://github.com/mayankshah1607/linkerd2-conformance#1-initial-bootstrapping)
+  - [Adding tests](https://github.com/mayankshah1607/linkerd2-conformance#2-adding-tests)
+  - [Wiring up newly added tests](https://github.com/mayankshah1607/linkerd2-conformance#3-wiring-up-newly-added-test)
+
+## Repository Structure
+
+- [`tests`](https://github.com/mayankshah1607/linkerd2-conformance/tree/master/specs) contains the tests for each of the features organized into separate packages
+- [`sonobuoy`](https://github.com/mayankshah1607/linkerd2-conformance/tree/master/sonobuoy) contains the items required to be able to run the tests as a [Sonobuoy](https://github.com/vmware-tanzu/sonobuoy) plugin
+- [`utils`](https://github.com/mayankshah1607/linkerd2-conformance/tree/master/utils) contains helper functions that can be used while writing conformance tests
+- [`bin`](https://github.com/mayankshah1607/linkerd2-conformance/blob/master/bin) contains useful helper scripts to build/push the Docker image and running the tests
+- [`test_template`](https://github.com/mayankshah1607/linkerd2-conformance/tree/master/test_template) contains a sample template for adding new tests under the `tests/` folder
```

Comment:
Does `ginkgo bootstrap` do the same thing?

Reply:
`ginkgo bootstrap` only provides us with a single file (`<package>_suite_test.go`) which acts as an entrypoint for the ginkgo test suite. Further, `ginkgo generate <testname>` gives us a single test file (`<testname>_test.go`) with a single spec.
The above process is a little naive in the sense that it will run the test cases generated in a random order (because thats how ginkgo deals with multiple `_test.go` files under a single package). We wanted more control over the randomization and better organization over our test cases, which is why we had to have a custom template instead.
[This section](https://github.com/mayankshah1607/linkerd2-conformance#adding-new-tests) explains how our test suites are organized and the significance of the `test_template` directory. :smile: 

Moreoever, each of our `_test.go` files need some logic in the `BeforeSuite` and `AfterSuite` blocks which does not come with `ginkgo generate`. Having a standard test template which satisfies all these requirements seemed a lot more effecient. :smile: